### PR TITLE
Add p-value indicator

### PR DESCRIPTION
**Pull request by tavinathanson**
In this PR:
- Add random cohort generation to make it easy to test out plots without using private data.
- Add a significance indicator that shows \* if <= 0.05, otherwise "ns", for both Fisher's and MW.

image

image

image

@arahuja

### CODE REVIEW

File: cohorts/plot.py [hunk 1]
```diff
@@ -22,7 +22,38 @@
 from sklearn.metrics import roc_curve
 from .model import bootstrap_auc
 
-def stripboxplot(x, y, data, ax=None, **kwargs):
+def vertical_percent(plot, percent=0.1):
+    plot_bottom, plot_top = plot.get_ylim()
+    return percent * (plot_top - plot_bottom)
+
+def set_min_y_from_data(y, data, plot):
+    """
```
Comment (arahuja) [inline cohorts/plot.py:43]: Does this need to take data and col name as an arg rather than the series itself?

File: cohorts/plot.py [hunk 2]
```diff
@@ -22,7 +22,38 @@
...
+    If data is >= 0, don't show negative axis values.
+    """
+    min_y_val = data[y].min()
+    if min_y_val >= 0:
+        # Add a bit of padding so that we don't cut the plot off in an ugly way
+        extra_space = -1 * vertical_percent(plot, 0.05)
+        plot.set_ylim(bottom=extra_space)
```
Comment (arahuja) [inline cohorts/plot.py]: What I've done sometimes is turn off the bottom and top most tick label to have the extra space but not display negatives.

```
    yticks = plot.yaxis.get_major_ticks()
    yticks[0].set_visible(False)
    yticks[-1].set_visible(False)
```

Setting the max might be nice too, since sometimes we want things between 0 - 100%

**Comment by arahuja** [comment]
Nice, like the random cohort. Only addition might making the significance level configurable, but not a must have.

File: cohorts/plot.py (hunk 2 above)
Comment (tavinathanson) [inline cohorts/plot.py]: Interesting; maybe that's a more stable approach? Not sure the best way to automatically set the max; hard to know if we're talking about 0 to 100 or not?

Comment (tavinathanson) [inline cohorts/plot.py]: What would `yticks[-1]` do if the tick is `-0.5`?

Comment (arahuja) [inline cohorts/plot.py]: Actually looking at this again, not 100% positive I know what's happening. Mind explaining in person?

File: cohorts/plot.py (hunk 1 above)
Comment (tavinathanson) [inline cohorts/plot.py:43]: Good point, fixed.

**Comment by tavinathanson** [comment]
@arahuja I'm now using `set_yticks` and `get_yticks` to filter negative ticks, and the same strategy for non-0 to 1 ticks in the case of Fisher's exact. I imagine we can generalize the latter in a later PR, as well as let 0.05 be customizable.

image

image

**Comment by tavinathanson** [comment]
@arahuja added a fix for the Fisher's exact ordering as well as custom styling in `styling.py`.

**Comment by tavinathanson** [comment]
image

**Comment by tavinathanson** [comment]
Also just simplified `__str__`s for `MannWhitneyResults` and Fisher's to not pollute notebooks.

**Comment by tavinathanson** [comment]
@arahuja sorry to keep adding to this PR, but also just made it easier (https://github.com/hammerlab/cohorts/pull/112/commits/0d70a60c71d25d912ed6fd2dacc948f92787c344) to deal with situations where I want to split on something other than `benefit` that isn't presently in the `DataFrame`: I can include it as another function in `on`, and then specify which function is the `plot_col` vs. the `boolean_col`:

Example:

```
cohort.plot_boolean(on={"Liver Met": lambda row: row["Liver Mets"] == 1, "Split": split_patients},
                    plot_col="Liver Met",
                    boolean_col="Split")
```

File: cohorts/plot.py [hunk 3]
```diff
@@ -45,6 +89,11 @@ def stripboxplot(x, y, data, ax=None, **kwargs):
         **kwargs
     )
 
+    hide_negative_y_ticks(plot)
```
Comment (arahuja) [inline cohorts/plot.py]: Are we always hiding negative ticks?

File: cohorts/plot.py [hunk 4]
```diff
@@ -82,11 +137,16 @@ def fishers_exact_plot(data, condition1, condition2, ax=None, alternative="two-s
...
-        data=data
+        data=data,
+        **kwargs
     )
+
     count_table = pd.crosstab(data[condition1], data[condition2])
     print(count_table)
     oddsratio, pvalue = fisher_exact(count_table, alternative=alternative)
+    hide_negative_y_ticks(plot)
+    only_percentage_ticks(plot)
```
Comment (arahuja) [inline cohorts/plot.py:150]: Are these two always used in tandem? Or does `only_percentage_ticks` cover `hide_negative_y_ticks`? If not, should it?

**Comment by arahuja** [comment]
Cool, the latest addition was something I would've found useful recently and may let me re-work how I did #111 to allow more customization. Will leave that off for now.

Left some general comments but makes sense if we want to get this in and come back to clean up some minor aspects.

**Comment by arahuja** [comment]
Ah another thing I just realized, I use `stripboxplot` on its own frequently, with no stat. test, might be useful (in the future) to not plot the line, or `None` as an option for `significance`

File: cohorts/plot.py (hunk 4 above)
Comment (tavinathanson) [inline cohorts/plot.py:150]: Great catch! Got lost in the fixes. Just re-added.

Comment (tavinathanson) [inline cohorts/plot.py:150]: Oops, that was intended for the above comment.

File: cohorts/plot.py (hunk 3 above)
Comment (tavinathanson) [inline cohorts/plot.py]: Great catch! Got lost in the fixes. Just re-added.

File: cohorts/plot.py (hunk 4 above)
Comment (tavinathanson) [inline cohorts/plot.py:150]: `only_percentage_ticks` is only used for Fisher's exact right now. Does that answer your question?

**Comment by tavinathanson** [comment]
I think all comments are addressed and will plan to merge on 🍏  unless I hear anything!

File: cohorts/plot.py (hunk 4 above)
Comment (arahuja) [inline cohorts/plot.py:150]: My question is:
- do you always need to call `hide_negative_y_ticks` before `only_percentage_ticks`?

If yes:
- should it just be called from `only_percentage_ticks`?
  If no:
- does `only_percentage_ticks` not remove negative ones on it's own?

Comment (tavinathanson) [inline cohorts/plot.py:150]: Oh. Right. Looks like that's confusingly named; perhaps I was intending them to be bundled but never did that, so `only_percentage_ticks` just handles <= 1. I just moved the Fisher's exact call to `hide_negative_y_ticks` into `only_percentage_ticks`.
